pkg/storage/opensearch.go

```diff
@@ -0,0 +1,385 @@
+				"term": map[string]any{fieldName: value},
+			})
+		}
+	}
+
+	if filter.ObserverType != "" {
+		addFilter(filter.ObserverType, osFieldMapping["observer_type"])
+	}
+	if filter.TargetType != "" {
+		addFilter(filter.TargetType, osFieldMapping["target_type"])
+	}
+	if filter.TargetID != "" {
+		addFilter(filter.TargetID, osFieldMapping["target_id"])
+	}
+	if filter.InitiatorType != "" {
+		addFilter(filter.InitiatorType, osFieldMapping["initiator_type"])
+	}
+	if filter.InitiatorID != "" {
+		addFilter(filter.InitiatorID, osFieldMapping["initiator_id"])
+	}
+	if filter.InitiatorName != "" {
+		addFilter(filter.InitiatorName, osFieldMapping["initiator_name"])
+	}
+	if filter.Action != "" {
+		addFilter(filter.Action, osFieldMapping["action"])
+	}
+	if filter.Outcome != "" {
+		addFilter(filter.Outcome, osFieldMapping["outcome"])
+	}
+	if filter.RequestPath != "" {
+		addFilter(filter.RequestPath, osFieldMapping["request_path"])
+	}
+
+	// Time range filters
+	if len(filter.Time) > 0 {
+		rangeQuery := map[string]any{osFieldMapping["time"]: map[string]any{}}
+		timeRange := rangeQuery[osFieldMapping["time"]].(map[string]any)
+
+		for key, value := range filter.Time {
+			switch key {
+			case "lt":
+				timeRange["lt"] = value
+			case "lte":
+				timeRange["lte"] = value
+			case "gt":
+				timeRange["gt"] = value
+			case "gte":
+				timeRange["gte"] = value
+			}
+		}
+
+		boolClause["filter"] = append(boolClause["filter"].([]any), map[string]any{"range": rangeQuery})
+	}
+
+	// Full-text search
+	if filter.Search != "" {
+		boolClause["must"] = append(boolClause["must"].([]any), map[string]any{
+			"query_string": map[string]any{
+				"query": filter.Search,
+			},
+		})
+	}
+
+	return boolQuery
+}
+
+// GetEvents grabs events for a given tenantID with filtering.
+func (os OpenSearch) GetEvents(ctx context.Context, filter *EventFilter, tenantID string) ([]*cadf.Event, int, error) {
+	index := indexName(tenantID)
+	logg.Debug("Looking for events in index %s", index)
+
+	// Build the query
+	query := buildBoolQuery(filter)
+
+	// Build the complete search body
+	searchBody := map[string]any{
+		"query": query,
+	}
+
+	// Add sorting
+	var sortArray []any
+	if filter.Sort != nil {
+		for _, fieldOrder := range filter.Sort {
+			sortOrder := "desc"
+			if fieldOrder.Order == "asc" {
+				sortOrder = "asc"
+			}
+			sortArray = append(sortArray, map[string]any{
+				osFieldMapping[fieldOrder.Fieldname]: map[string]any{"order": sortOrder},
+			})
+		}
+	}
+
+	// Always sort by time descending as default
+	sortArray = append(sortArray, map[string]any{
+		osFieldMapping["time"]: map[string]any{"order": "desc"},
+	})
+	searchBody["sort"] = sortArray
+
+	// Add pagination
+	offset := min(filter.Offset, math.MaxInt32)
+	limit := min(filter.Limit, math.MaxInt32)
+	searchBody["from"] = offset
+	searchBody["size"] = limit
+
+	// Convert to JSON
+	bodyJSON, err := json.Marshal(searchBody)
+	if err != nil {
+		return nil, 0, err
```

Comment:
```suggestion
		return []*cadf.Event, 0, err
```

I would return an empty list and not nil here. It would make it easier for parsing by the end user/frontend/secondary step. Otherwise there exists a risk of a nilpointer would the user not have made proper nil-assertions. As the user expects a [] this would be a likely error downstream.

Reply:
I appreciate the suggestion, but the current pattern is to return nil, and check for errors before using the slice. Changing this pattern is an idea perhaps worth persuing, but I'm concerned about scope creep for this PR. This PR is trying to replace the unmaintained library used before. I'm okay if we make a separate PR to change our error handling pattern, but I don't think we should do it in this pr. Perhaps I'm wrong though, I'm happy if you want to push back and argue that we should do it anyway.